src/entities/createActionListenerMiddleware.ts

```diff
@@ -0,0 +1,268 @@
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any>,
+    options?: ActionListenerOptions<AnyAction, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type
+      },
+      meta: {
+        listener,
+        options
+      }
+    }
+  }
+) as BaseActionCreator<
+  { type: string },
+  'actionListenerMiddleware/add',
+  {
+    listener: ActionListener<any, any, any>
+    options: ActionListenerOptions<any, any, any>
+  }
+> & {
+  <C extends TypedActionCreator<any>, S, D extends Dispatch>(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D>,
+    options?: ActionListenerOptions<ReturnType<C>, S, D>
+  ): AddListenerAction<ReturnType<C>, S, D>
+
+  <S, D extends Dispatch>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D>,
+    options?: ActionListenerOptions<AnyAction, S, D>
+  ): AddListenerAction<AnyAction, S, D>
+}
+
+interface RemoveListenerAction<
+  A extends AnyAction,
+  S,
+  D extends Dispatch<AnyAction>
+> {
+  type: 'actionListenerMiddleware/remove'
+  payload: {
+    type: string
+  }
+  meta: {
+    listener: ActionListener<A, S, D>
+  }
+}
+
+export const removeListenerAction = createAction(
+  'actionListenerMiddleware/remove',
+  function prepare(
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type
+      },
+      meta: {
+        listener
+      }
+    }
+  }
+) as BaseActionCreator<
+  { type: string },
+  'actionListenerMiddleware/remove',
+  {
+    listener: ActionListener<any, any, any>
+  }
+> & {
+  <C extends TypedActionCreator<any>, S, D extends Dispatch>(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D>
+  ): RemoveListenerAction<ReturnType<C>, S, D>
+
+  <S, D extends Dispatch>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D>
+  ): RemoveListenerAction<AnyAction, S, D>
+}
+
+export function createActionListenerMiddleware<
+  S,
+  D extends Dispatch<AnyAction> = Dispatch
+>() {
+  type ListenerEntry = ActionListenerOptions<any, S, D> & {
+    listener: ActionListener<any, S, D>
+  }
+
+  const listenerMap: Record<string, Set<ListenerEntry>> = {}
+  const middleware: Middleware<
+    {
+      (action: Action<'actionListenerMiddleware/add'>): Unsubscribe
+    },
+    S,
+    D
+  > = api => next => action => {
+    if (addListenerAction.match(action)) {
+      const unsubscribe = addListener(
+        action.payload.type,
+        action.meta.listener,
+        action.meta.options
+      )
+      delete action.meta
+      next(action)
```

Comment:
Would we actually want to forward the `addListener` action onwards at all?

Reply:
I do have something in the back of my mind that someone talked about maybe wanting to track all active subscriptions in a state or so - in that case it would definitely be useful. On the other hand, the `delete action.meta` would be counter-productive then.

Also, if it is passed thorugh, it can be observed in the devTools, wich might be nice.

Reply:
Yeah, [that was me](https://github.com/reduxjs/redux-toolkit/issues/237#issuecomment-590269935) (back when I still thought I'd have the time to look at this 😓), although I did concede that in practise, nothing ever actually reacted to those events.  The observability in the dev-tools has saved us on multiple occasions, so I'd advocate for keeping it.

The only thing I'd recommend is cleansing the action of any functions so that it serialisable by the time it hit the store. I thought RTK actually raised a warning if the action wasn't?

Reply:
Just noticed the the `delete action.meta`.

Not sure if you want to modify the original action in case the dispatcher is reusing the same action (i.e. if they're adding and removing listeners on component mount/dismount).

Reply:
Is there really a meaningful benefit to having actions like "add/remove listener" show up in the DevTools? I've never been a fan of the idea of dispatching actions _just_ to get a logging effect.

Reply:
Honestly, 90% of the time I look into the devtools to just look at the action names (sometimes payloads) to get a feeling for the flow what happened.
10% for state changes. Never use time travel.
So I'm very much in favor of forwarding it, but I agree that it has to have some use.

Right now I'm thinking about preventing the "listen" action from propagating and dispatching a new "started_listening" action, which might also contain a unique id that could be used to unsubscribe that listener.

Reply:
Hmm. While trying to implement this "started_listening", it seems non-trivial:
* do we dispatch that action only when it is subscribed via `addListener` action?
* if we also dispatch it, what if the middleware isn't used by a store yet - or used by multiple stores?
* would we dispatch something different if a listener was not added because it was already in use? Same for removal.

So for now, I'm tending towards cutting that feature altogether. Let's get this out as simple as possible, we can always add it back later on, with some separate discussions. Right now it looks like it might not be worth the amount of work that would be required.

Reply:
Removed it for now. Let's re-evaluate that later, in a separate issue.